data/posts/fog-of-war.md

```diff
@@ -0,0 +1,206 @@
+- **Route implementation**: Parts of the route needed to load data and render the UI (`loader`, `Component`, `ErrorBoundary`, etc.)
+
+In order to implement optimizations mentioned in the previous section, Remix needs to know all of your route _definitions_ in the client so that it can match routes based on nothing more than a `<Link to>`. Once a link is clicked and routes are matched, Remix can fetch data and download the route _implementations_ in parallel.
+
+To do this, Remix ships a **Route Manifest** to the client that contains all of your route definitions plus a small amount of metadata. This manifest allows Remix to build a client-side route tree without including the underlying route implementations. Those implementations are loaded via [route.lazy][route-lazy] when a route is navigated to.
+
+Here's a sample route manifest entry for the root route on https://remix.run:
+
+```json
+"root": {
+  "id": "root",
+  "path": "",
+  "hasAction": false,
+  "hasLoader": true,
+  "hasClientAction": false,
+  "hasClientLoader": false,
+  "hasErrorBoundary": true,
+  "module": "/assets/root-x1zXK6d6.js",
+  "imports": [
+    "/assets/entry.client-uo5Ucqv5.js",
+    "/assets/utils-c6MmN8mv.js",
+    "/assets/color-scheme-y8FjcTs4.js",
+    "/assets/icons-6-7TeyZS.js"
+  ],
+  "css": [
+    "/assets/root-Wq_jPy7B.css"
+  ]
+}
+```
+
+Over time, as the user navigates around—more and more route implementations are downloaded via `route.lazy` and the route tree becomes more and more complete.
+
+This approach works for most applications - since the manifest is pretty lightweight and compresses well since it's a repetitive key/value JSON structure. For example, the manifest for https://remix.run/ contains 50 routes, weighs 19.6Kb uncompressed, but only sends 2.6Kb over the wire after compression.
+
+However, Remix doesn't only want to to provide good performance for small-to-medium sized applications—we want _large_ and _extremely large_ applications to be fast by default too! Because we love dog-fooding React Router and Remix on the myriad of applications at Shopify, both internal and public-facing, we knew our current strategy was not cutting it for those larger applications.
+
+As we began rolling out Remix to https://www.shopify.com, we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more)–the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module, and thus duplicating the module information (it's path, it's other module `imports`, etc.). This resulted in a manifest that was ~85Kb compressed, and ~10Mb uncompressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.
```

Comment:
```suggestion
As we began rolling out Remix to https://www.shopify.com, we realized _just how big_ that site is. When you take into account all of the routes and their internationalized URLs (i.e., `/pricing`, `/en/pricing`, `/es/precios` and many more)–the app had over 1300 routes! And because Remix doesn't have a good solution for URL aliasing (yet!), many of the route entries were duplicates pointing to the same route module, and thus duplicating the module information (its path, its other module `imports`, etc.). This resulted in a manifest that was ~85Kb compressed, and ~10Mb uncompressed. On slower devices, this could have a noticeable impact on page load times for the device to decompress/parse/compile/execute the JS module.
```